arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -0,0 +1,1563 @@
+          builder.add("toServer", VPackValue(ctx->toServer));
+          builder.add("isLeader", VPackValue(isLeader));
+          builder.add("remainsFollower", VPackValue(isLeader));
+          builder.add("creator", VPackValue(ServerState::instance()->getId()));
+          builder.add("timeCreated", VPackValue(timepointToString(std::chrono::system_clock::now())));
+        })
+        .done()
+      .done();
+  }
+
+  auto self(shared_from_this());
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx)).thenValue(
+    [this, ctx = std::move(ctx), jobId = std::move(jobId)](AsyncAgencyCommResult &&result) {
+
+      if (result.ok() && result.statusCode() == fuerte::StatusOK) {
+        VPackBuffer<uint8_t> payload;
+        {
+          VPackBuilder builder(payload);
+          VPackObjectBuilder ob(&builder);
+          builder.add(StaticStrings::Error, VPackValue(false));
+          builder.add("code", VPackValue(int(ResponseCode::ACCEPTED)));
+          builder.add("id", VPackValue(jobId));
+        }
+
+        resetResponse(rest::ResponseCode::ACCEPTED);
+        response()->setPayload(std::move(payload), true);
+
+      } else {
+        generateError(result.asResult());
+      }
+    });
+}
+
+RestStatus RestAdminClusterHandler::handlePostMoveShard(std::unique_ptr<MoveShardContext>&& ctx) {
+  std::shared_ptr<LogicalCollection> collection = server().getFeature<ClusterFeature>()
+    .clusterInfo().getCollectionNT(ctx->database, ctx->collection);
+
+  if (collection == nullptr){
+    generateError(ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND,
+      "database/collection not found");
+    return RestStatus::DONE;
+  }
+
+  ctx->collectionID = TRI_RidToString(collection->planId());
+  auto planPath = arangodb::cluster::paths::root()->arango()->plan();
+
+  VPackBuffer<uint8_t> trx;
+  {
+    VPackBuilder builder(trx);
+    arangodb::agency::envelope<VPackBuilder>::create(builder).read()
+      .key(planPath->dBServers()->str())
+      .key(planPath->collections()->database(ctx->database)
+        ->collection(ctx->collectionID)->str())
+      .done()
+    .done();
+  }
+
+  // gather information about that shard
+  return waitForFuture(AsyncAgencyComm().getValues(planPath).thenValue(
+      [this, ctx = std::move(ctx)](AgencyReadResult &&result) mutable {
+
+        if (result.ok()) {
+          switch (result.statusCode()) {
+            case fuerte::StatusOK:
+              return createMoveShard(std::move(ctx), result.slice().at(0));
+            case fuerte::StatusNotFound:
+              generateError(rest::ResponseCode::NOT_FOUND, TRI_ERROR_HTTP_NOT_FOUND, "unknown collection");
+              return futures::makeFuture();
+            default:
+              break;
+          }
+        }
+
+        // to some checks that this shard is moveable
+        generateError(result.asResult());
+        return futures::makeFuture();
+    }).thenError<VPackException>([this](VPackException const& e) {
```

Comment:
do we need to capture `self` here?

Reply:
This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.